06x-new-kafka-roller.md

```diff
@@ -0,0 +1,325 @@
+| maxRestartAttempts     | 3             | No              | The maximum number of times a node can be restarted before failing the reconciliation. This is checked against the node's `numRestartAttempts`.                                                                                                               |
+| maxReconfigAttempts    | 3             | No              | The maximum number of times a node can be dynamically reconfigured before restarting it. This is checked against the node's `numReconfigAttempts`.                                                                                                                        |
+| maxRetries             | 10            | No              | The maximum number of times a node can be retried after not meeting the safety conditions e.g. availability check failed.  This is checked against the node's `numRetries`.                                                                                                                      |
+| operationTimeoutMs | 60 seconds    | Yes             | The maximum amount of time we will wait for nodes to transition to `READY` state after an operation in each retry. This is already exposed to the user via environment variable `STRIMZI_OPERATION_TIMEOUT_MS`. |
+| maxRestartParallelism  | 1             | Yes             | The maximum number of broker nodes that can be restarted in parallel. This will be exposed to the user via the new environment variable `STRIMZI_MAX_RESTART_BATCH_SIZE`.     
+| postRestartDelay  | 0             | Yes             | Delay between restarts of nodes or batches. It's set to 0 by default, but can be adjusted by users to slow down the restarts. This will also help JIT to reach a steady state and to reduce impact on clients.   
+| restartAndPreferredLeaderElectionDelay  | 10 seconds             | No             | Delay between restart and triggering partition leader election so that just-rolled broker is leading all the partitions it is the preferred leader for. This is to avoid situations where leaders moving to a newly started node that does not yet have established networking to some outside networks, e.g. through load balancers.
+
+### Algorithm
+
+1. **Initialize Context for Each Node:**
+   Create a context object with the following data:
+   ```
+   Context: {
+       nodeRef: <NodeRef from KafkaReconciler>,
+       nodeRoles: <Set using pod labels `strimzi.io/controller-role` and `strimzi.io/broker-role`>,
+       state: UNKNOWN,
+       lastTransition: <SYSTEM_TIME>,
+       reason: <Result of predicate function from KafkaReconciler>,
+       numRestartAttempts: 0,
+       numReconfigAttempts: 0,
+       numRetries: 0
+   }
+   ```
+   Contexts are recreated in each reconciliation with the above initial data.
+
+2. **Transition Node States:**
+   Update each node's state based on information from abstracted sources. If failed to retrieve information, the reconciliation fails and restarts from step 1.
```

Comment:
I would say that this is about "loading" or "building" the current state of the node. Usually in our FSMs (i.e. rebalancing), this state is coming from a custom resource, here it's coming from different sources. Maybe we can say it better instead of "update each node's ...." .

Reply:
I'm not sure what would be better, does `Load each node's state based on information...` sound better?

We build a context in step 1, which has state UNKNOWN. In this step, we are updating this state based on the information from the sources.  So to me, `update each node's state` sounds fine.